core/src/main/java/com/datastrato/gravitino/storage/relational/service/FilesetMetaService.java

```diff
@@ -213,6 +218,54 @@ public boolean deleteFileset(NameIdentifier identifier) {
     return true;
   }
 
+  public int deleteFilesetAndVersionMetasByLegacyTimeLine(Long legacyTimeLine, int limit) {
+    int filesetDeletedCount =
+        SessionUtils.doWithCommitAndFetchResult(
+            FilesetMetaMapper.class,
+            mapper -> {
+              return mapper.deleteFilesetMetasByLegacyTimeLine(legacyTimeLine, limit);
+            });
+    int filesetVersionDeletedCount =
+        SessionUtils.doWithCommitAndFetchResult(
+            FilesetVersionMapper.class,
+            mapper -> {
+              return mapper.deleteFilesetVersionsByLegacyTimeLine(legacyTimeLine, limit);
+            });
+    return filesetDeletedCount + filesetVersionDeletedCount;
+  }
+
+  public int deleteFilesetVersionsByRetentionCount(Long versionRetentionCount, int limit) {
+    // get the current version of all filesets.
+    List<ImmutablePair<Long, Integer>> filesetCurVersions =
+        SessionUtils.getWithoutCommit(
+            FilesetVersionMapper.class,
+            mapper -> mapper.selectFilesetVersionsByRetentionCount(versionRetentionCount));
+
+    // soft delete old versions that are older than or equal to (currentVersion -
+    // versionRetentionCount).
+    int totalDeletedCount = 0;
+    for (ImmutablePair<Long, Integer> filesetCurVersion : filesetCurVersions) {
+      long versionRetentionLine = filesetCurVersion.getValue() - versionRetentionCount;
+      int deletedCount =
+          SessionUtils.doWithCommitAndFetchResult(
+              FilesetVersionMapper.class,
+              mapper ->
+                  mapper.softDeleteFilesetVersionsByRetentionLine(
```

Comment:
I'm curious as to why we use soft delete here.

Reply:
Actually, both soft and hard deletions are ok here. Soft deletion is preferred mainly to keep the logic unified, that is, all scene deletions are first soft deletion, then hard deletion. Do you have any suggestions?

Reply:
Should we need another logic to physically drop the version data? or does the version data need to be kept forever?

Just curious, thought current code will not affect normal read/write operation.

Reply:
Yes, another logic `hardDeleteLegacyData` will physically delete it, and the unit test has covered this.

Reply:
OK